### PR TITLE
[feature](auth)Implementing privilege management with rbac model

### PR DESCRIPTION
**Pull request by zddr**
# Proposed changes

Issue Number: close #xxx

## Problem summary

https://cwiki.apache.org/confluence/display/DORIS/DSIP-030%3A+RBAC

change implement of auth to rbac

each user has one default role which can not be drop;

if you grant priv to user,it will grant to default role ,

In the current pr, the user can still only have one role other than the default role, but in the future, the user and role will be many-to-many

rename PaloRole,PaloAuth,PaloPrivilege to Role,Auth,Privilege

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [x] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.3 seconds
 load time: 475 seconds
 storage size: 17170642077 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230203054345_clickbench_pr_89907.html

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java [hunk 1]
```diff
@@ -40,19 +44,25 @@
...
 
-    public PaloRole getRole(String role) {
-        return roles.get(role);
+    public Role getRole(String name) {
+        return roles.get(name);
     }
 
-    public PaloRole addRole(PaloRole newRole, boolean errOnExist) throws DdlException {
-        PaloRole existingRole = roles.get(newRole.getRoleName());
+    public Role addRole(Role newRole, boolean errOnExist) throws DdlException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java:64]: How about name it as `addOrMergeRole`?

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+
+    public static Role OPERATOR = new Role(OPERATOR_ROLE,
+            TablePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV),
+            ResourcePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV));
+    public static Role ADMIN = new Role(ADMIN_ROLE,
+            TablePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV),
+            ResourcePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV));
+
+    private String roleName;
+    private Map<TablePattern, PrivBitSet> tblPatternToPrivs = Maps.newConcurrentMap();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java:60]: Add comment to explain which field will be persist

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java [hunk 3]
```diff
@@ -72,37 +61,29 @@ public boolean hasPriv(String host, String user, PrivPredicate wanted) {
...
-                    existingEntry.setPrivSet(newEntry.getPrivSet());
-                    existingEntry.setSetByDomainResolver(newEntry.isSetByDomainResolver());
-                    LOG.debug("reset priv entry: {}", existingEntry);
-                } else if (!newEntry.isSetByDomainResolver()) {
-                    mergePriv(existingEntry, newEntry);
-                    existingEntry.setSetByDomainResolver(false);
-                    LOG.debug("merge priv entry: {}", existingEntry);
-                }
+                mergePriv(existingEntry, newEntry);
+                LOG.debug("merge priv entry: {}", existingEntry);
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java:80]: 1. create user user1@domain   -> domain(ip1, ip2)
2. create user user1@ip1
How to handle it?

[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java:80]: check this `checkOperationAllowed` logic

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Auth.java [hunk 4]
```diff
@@ -0,0 +1,1627 @@
...
+                    if (!user.isSetByDomainResolver()) {
+                        TablePrivTable tablePrivTable = getUserTblPrivTable(user.getUserIdentity());
+                        if (tablePrivTable.isEmpty()) {
+                            continue;
+                        }
+                        for (PrivEntry entry : tablePrivTable.getEntries()) {
+                            TablePrivEntry tablePrivEntry = (TablePrivEntry) entry;
+                            String dbName = ClusterNamespace.getNameFromFullName(tablePrivEntry.getOrigDb());
+                            String tblName = tablePrivEntry.getOrigTbl();
+                            if (dbName.equals("information_schema" /* Don't show privileges in information_schema */)
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Auth.java:1248]: InfoschemaDb.xxxx

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/GlobalPrivEntry.java [hunk 5]
```diff
@@ -18,144 +18,52 @@
...
-    private byte[] password;
-    // set domainUserIdent when this a password entry and is set by domain resolver.
-    // so that when user checking password with user@'IP' and match a entry set by the resolver,
-    // it should return this domainUserIdent as "current user". And user can use this user ident to get privileges
-    // further.
-    private UserIdentity domainUserIdent;
+    @Deprecated
+    protected byte[] password;
+    @Deprecated
+    protected UserIdentity domainUserIdent;
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/GlobalPrivEntry.java:34]: There is also privSet inherited from the parent class. The reason why the parent class cannot be used directly is that 1. The parent class is abstract, 2. It is not intuitive

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivEntry.java [hunk 6]
```diff
@@ -82,108 +112,7 @@ protected PrivEntry(PatternMatcher hostPattern, String origHost, PatternMatcher
...
-     *      2. A's self members
-     *      3. B's self members
-     *      4. C's self members
-     *
-     *  Which is same as Write order.
-     */
+    @Deprecated
     public static PrivEntry read(DataInput in) throws IOException {
         String className = Text.readString(in);
         if (className.startsWith("com.baidu.palo")) {
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivEntry.java:118]: and remove same in PrivTable

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java [hunk 7]
```diff
@@ -40,19 +44,25 @@
 import java.util.stream.Stream;
 
 public class RoleManager implements Writable {
-    private Map<String, PaloRole> roles = Maps.newHashMap();
+    private static final Logger LOG = LogManager.getLogger(RoleManager.class);
+    //prefix of each user default role
+    public static String DEFAULT_ROLE_PREFIX = "default_role_rbac_";
+
+    private Map<String, Role> roles = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java:51]: Concurrency control is delegated by Auth, so not concurrentMap

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserManager.java [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+public class UserManager implements Writable {
+    public static final String ANY_HOST = "%";
+    private static final Logger LOG = LogManager.getLogger(UserManager.class);
+    //    Set<User> users = new HashSet<>();
+    //One name may have multiple User,because host can be different
+    private Map<String, List<User>> nameToUsers = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserManager.java:55]: // Concurrency control is delegated by Auth, so not concurrentMap

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserManager.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        }
+        if (CollectionUtils.isEmpty(nameToLists)) {
+            nameToUsers.remove(userIdent.getQualifiedUser());
+        } else {
+            Collections.sort(nameToLists);
+        }
+    }
+
+    public Map<String, List<User>> getNameToUsers() {
+        return nameToUsers;
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserManager.java:247]: Solve security issues by limiting Auth exposure to UserManager

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserRoleManager.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+public class UserRoleManager implements Writable {
+    private Map<UserIdentity, Set<String>> userToRoles = Maps.newConcurrentMap();
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserRoleManager.java:37]: Concurrency control is delegated by Auth, so not concurrentMap

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java:80]: "checkOperationAllowed" ===>now,PrivTable only related with 'role',do not exist this logic

[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java:80]: create user user1@domain -> domain(ip1, ip2)
create user user1@ip1
How to handle it?
===================>
update user1@ip1.setByResolver to false

[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java:80]: How to handle domain now?
=====================>
now,domainUser,ipUser,and resolveUserByDomain all manage By userManager,The logic of scheduled tasks is the same as before

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserRoleManager.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    public static UserRoleManager read(DataInput in) throws IOException {
+        String json = Text.readString(in);
+        return GsonUtils.GSON.fromJson(json, UserRoleManager.class);
+    }
+
+    @Override
+    public void gsonPostProcess() throws IOException {
+        for (Entry<UserIdentity, Set<String>> entry : userToRoles.entrySet()) {
+            for (String roleName : entry.getValue()) {
+                Set<UserIdentity> userIdentities = roleToUsers.get(roleName);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserRoleManager.java:124]: The `roleToUsers` maybe null here

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+            }
+            return role;
+        } else {
+            String json = Text.readString(in);
+            return GsonUtils.GSON.fromJson(json, Role.class);
+        }
+    }
+
+
+    public void readFields(DataInput in) throws IOException, DdlException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java:564]: use `private` and mark it as @deprecated

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java [hunk 13]
```diff
@@ -184,8 +167,8 @@ public static RoleManager read(DataInput in) throws IOException {
     public void write(DataOutput out) throws IOException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/RoleManager.java:167]: Change to Gson to, use meta version to make it compatible

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Privilege.java [hunk 14]
```diff
@@ -44,8 +44,8 @@ public enum PaloPrivilege {
             USAGE_PRIV
     };
 
-    public static Map<PaloPrivilege, String> privInPaloToMysql =
-            ImmutableMap.<PaloPrivilege, String>builder() // No NODE_PRIV and ADMIN_PRIV in the mysql
+    public static Map<Privilege, String> privInPaloToMysql =
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Privilege.java:47]:
```suggestion
    public static Map<Privilege, String> privInDorisToMysql =
```

**Comment by morningman** [comment]
And need to check that the user name can not start with `DEFAULT_ROLE_PREFIX`

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Auth.java [hunk 15]
```diff
@@ -0,0 +1,1644 @@
...
+        return checkTblPriv(ctx.getCurrentUserIdentity(), qualifiedCtl, qualifiedDb, tbl, wanted);
+    }
+
+    public boolean checkTblPriv(ConnectContext ctx, String qualifiedDb, String tbl, PrivPredicate wanted) {
+        return checkTblPriv(ctx, DEFAULT_CATALOG, qualifiedDb, tbl, wanted);
+    }
+
+    public boolean checkTblPriv(ConnectContext ctx, TableName tableName, PrivPredicate wanted) {
+        Preconditions.checkState(tableName.isFullyQualified());
+        return checkTblPriv(ctx, tableName.getDb(), tableName.getTbl(), wanted);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Auth.java:323]:
```suggestion
        return checkTblPriv(ctx, tableName.getCtl(), tableName.getDb(), tableName.getTbl(), wanted);
```

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**